packages/core/src/queue/event-queue.ts

```diff
@@ -0,0 +1,319 @@
+        if (plugin.type === 'destination') {
+          this.failedInitializations.push(plugin.name)
+          console.warn(plugin.name, err)
+
+          ctx.log('warn', 'Failed to load destination', {
+            plugin: plugin.name,
+            error: err,
+          })
+
+          return
+        }
+
+        throw err
+      })
+  }
+
+  async deregister(
+    ctx: CoreContext,
+    plugin: CorePlugin,
+    instance: CoreAnalytics
+  ): Promise<void> {
+    try {
+      if (plugin.unload) {
+        await Promise.resolve(plugin.unload(ctx, instance))
+      }
+
+      this.plugins = this.plugins.filter((p) => p.name !== plugin.name)
+    } catch (e) {
+      ctx.log('warn', 'Failed to unload destination', {
+        plugin: plugin.name,
+        error: e,
+      })
+    }
+  }
+
+  async dispatch(ctx: CoreContext): Promise<CoreContext> {
+    ctx.log('debug', 'Dispatching')
+    this.emit('message_dispatched')
+    ctx.stats?.increment('message_dispatched')
+
+    this.queue.push(ctx)
+    const willDeliver = this.subscribeToDelivery(ctx)
+    this.scheduleFlush(0)
+    return willDeliver
+  }
+
+  private async subscribeToDelivery(ctx: CoreContext): Promise<CoreContext> {
+    return new Promise((resolve) => {
+      const onDeliver = (flushed: CoreContext, delivered: boolean): void => {
+        if (flushed.isSame(ctx)) {
+          this.off('flush', onDeliver)
+          if (delivered) {
+            resolve(flushed)
+          } else {
+            resolve(flushed)
+          }
+        }
+      }
+
+      this.on('flush', onDeliver)
+    })
+  }
+
+  async dispatchSingle(ctx: CoreContext): Promise<CoreContext> {
+    ctx.log('debug', 'Dispatching')
+    this.emit('message_dispatched')
+
+    this.queue.updateAttempts(ctx)
+    ctx.attempts = 1
+
+    return this.deliver(ctx).catch((err) => {
+      if (err instanceof ContextCancelation && err.retry === false) {
+        ctx.setFailedDelivery({ reason: err })
+        return ctx
+      }
+
+      const accepted = this.enqueuRetry(err, ctx)
+      if (!accepted) {
+        ctx.setFailedDelivery({ reason: err })
+        return ctx
+      }
+
+      return this.subscribeToDelivery(ctx)
+    })
+  }
+
+  isEmpty(): boolean {
+    return this.queue.length === 0
+  }
+
+  private scheduleFlush(timeout = 500): void {
+    if (this.flushing) {
+      return
+    }
+
+    this.flushing = true
+
+    setTimeout(() => {
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.flush().then(() => {
+        setTimeout(() => {
+          this.flushing = false
+
+          if (this.queue.length) {
+            this.scheduleFlush(0)
+          }
+        }, 0)
+      })
+    }, timeout)
+  }
+
+  private async deliver(ctx: CoreContext): Promise<CoreContext> {
+    await this.criticalTasks.done()
+
+    const start = Date.now()
+    try {
+      ctx = await this.flushOne(ctx)
+      const done = Date.now() - start
+      this.emit('delivery_success', done)
```

Comment:
new emitter -- I could see us moving the stats stuff outside of this function altogether and just regally on emitter events

Reply:
/cc @zikaari